Fix/relax types

faxm0dem opened the pull request:
We just converted our 632 patterns to the new v5 format.
We ran into a few incompatible changes, and this PR allows all our patterns to work out.

Basically we relax a few type constraints, which are a bit too tight compared to what's allowed in patterndb

faxm0dem committed:
Allow example values to be empty

In patterndb, an empty string is a valid one

faxm0dem committed:
Allow a rule's ID to be empty

Given `String[1] $id = $title` the rules will always have an id.

faxm0dem committed:
Allow program to be empty in an example

faxm0dem committed:
Apparently, there's a null type

```
Wrong match name='apache.query_param', value='', type='null', expected='', expected_type='string'
```

smortex commented on types/example.pp:
```diff
@@ -1,14 +1,14 @@
 type Patterndb::Example = Struct[
   {
-    program      => String[1],
+    program      => String,
```
/me reads xsd :exploding_head: 

If I am correct, the schema says that *the attribute is optional* but says nothing about *its length*.  I tried to remove the program from one of my examples, and validation pass but the test ignore the example I modified:

```
NOT Testing message as program is unset: ...
```

Of course, setting program to an empty string fail validation, which I expected.


If we want to allow empty string, I am in favor of being explicit about it: `String[0]` (After some years of work in numerous modules, I came to the conclusion that I should consider all usage of `String` without length as a code smell :smile:)

If we want to stick to the XML schema, we can also make this optional (will write about it bellow), but as far as I am concerned, it feels broken.

faxm0dem replied:
FTR leaving the program out is equivalent to a wildcard for all programs : the rule will match all programs

smortex commented on types/example.pp:
```diff
@@ -1,14 +1,14 @@
           {
-            type  => Optional[Enum['datetime', 'double', 'integer', 'json', 'string']],
-            value => String[1],
+            type  => Optional[Enum['datetime', 'double', 'integer', 'json', 'string', 'null']],
```
The XML schema does not list the allowed values so this list was the one from the comment. 'null' does not appear there, but indeed, when a field is absent, the test match its current value as `` and its current type as `null`, so I guess it make sense to add it.

While changing it, I think we should change the `Optional` location which is not exactly what I expected: with this syntax we allow an undef for "type", but in reality, "type" is optional and if set should be one of the expected values:

```puppet
type Foo = Struct[
  {
    foo => Optional[String[1]],
  },
]

warning({} =~ Foo)               # true
warning({foo => ""} =~ Foo)      # false
warning({foo => "value"} =~ Foo) # true
warning({foo => undef } =~ Foo)  # true <---

type Bar = Struct[
  {
    Optional[foo] => String[1],
  },
]

warning({} =~ Bar)               # true
warning({foo => ""} =~ Bar)      # false
warning({foo => "value"} =~ Bar) # true
warning({foo => undef } =~ Bar)  # false <---

```

So I suggest:
```suggestion
            Optional[type]  => Enum['datetime', 'double', 'integer', 'json', 'string', 'null'],
```

faxm0dem replied:
Wait, `warning({} =~ Foo)               # true` ? Isn't that a bug ?

smortex commented on types/example.pp:
```diff
@@ -1,14 +1,14 @@
-            type  => Optional[Enum['datetime', 'double', 'integer', 'json', 'string']],
-            value => String[1],
+            type  => Optional[Enum['datetime', 'double', 'integer', 'json', 'string', 'null']],
+            value => String,
```
```suggestion
            value => String[0],
```

faxm0dem replied:
I agree it's more concise, but both forms are functionally [identical](https://www.puppet.com/docs/puppet/7/lang_data_string#lang_data_string_data_type)

smortex replied:
I think at some point I will hack a puppet-lint plugin to find these in the puppetlabs / voxpupuli modules.  The fact that (most) String already have a lower bound here when required mean that it will be easy for us to fix those if needed in the future.

smortex commented on types/rule.pp:
```diff
@@ -1,7 +1,7 @@
   {
     patterns        => Array[String[1]],
-    id              => String[1],
+    id              => Optional[String[1]],
```
I guess that all the Struct I have ever created are wrong and should be fixed from `foo => Optional[bar]` to `Optional[foo] => bar` :unamused:

faxm0dem replied:
I replaced all old-wrong-old-style optionals

smortex replied:
Awesome!

smortex reviewed (commented):
I added some in-line comments for minor things.  The CI failures on Debian are due to https://github.com/syslog-ng/syslog-ng/issues/4585 but there was no release since that so we can ignore them for now.

faxm0dem committed:
Make type explicitly optional

This makes the following invalid (yes, surprisingly it was valid before) :

```
{ 'type' => undef }
```

Co-authored-by: Romain Tartière <romain@blogreen.org>

smortex commented on types/example.pp:
```diff
@@ -1,14 +1,14 @@
           {
-            type  => Optional[Enum['datetime', 'double', 'integer', 'json', 'string']],
-            value => String[1],
+            Optional[type]  => Enum['datetime', 'double', 'integer', 'json', 'string', 'null'],
```
`type` is a keyword, and seems to confuse the parser now :upside_down_face:.

This will work:

```suggestion
            Optional['type'] => Enum['datetime', 'double', 'integer', 'json', 'string', 'null'],
```

faxm0dem replied:
I replaced all wrong-old-style-optionals in latest commit

faxm0dem committed:
Fix non-mandatory options

faxm0dem committed:
Fix bareword

Co-authored-by: Romain Tartière <romain@blogreen.org>

faxm0dem commented:
Something very annoying also, is that `value` must be a String.
This is awkward:

```
test_values:
  i_am_a_number: '123'
  type: integer
```

This feels more right:

```
test_values:
  i_am_a_number: 123
  type: integer
```

Should we do `Variant[String,Number]` ?

faxm0dem committed:
Allow for numbers in example value

smortex commented:
> Should we do `Variant[String,Numeric]` ?

We can.

In this case we can probably also add `Float` to that list and assume the end-users will only pass literals and not compute the values (`0.02` vs `0.1 * 0.2`)?

smortex reviewed (approved):
This looks good.  You suggested relaxing some data types which I am fine with if you want to do it.

You can also rebase this branch on top of master so that CI check it does not break something unexpectedly.

Anyway, LGTM!

faxm0dem commented:
I *think* we need to resolve pull/44 first as this one depends on (parts of) it

smortex commented:
I updated #44 and it is ready for another review/merging.  I see you cherry-picked my commit in this branch, I guess that removing that commit and rebasing on top of master will resolve conflicts.

faxm0dem committed:
Don't log output if merge/deploy/test works

faxm0dem commented:
Done
